sage.py

```diff
@@ -349,125 +279,132 @@ def load_data(path, t):
-            bot = np.add(t[0], t[1])
-            for i in inde[2:]:
-                bot = np.add(bot, t[i]).tolist()
-            plot = plt.bar(ind, t[tid], width,
-                 bottom=bot) 
-        p.append(plot)
+            index = [x for x in range(tid)]
+            bottom = np.add(percentages[0], percentages[1])
+            for i in index[2:]:
+                bottom = np.add(bottom, percentages[i]).tolist()
+            plot = plt.bar(indices, percentages[tid], bar_width, bottom=bottom)
+        plots.append(plot)
 
         # TODO: Decide whether to put it like this or normalize over columns
-    #print(t)
-    plt.ylabel('Percentage of occurance')
+    plt.ylabel('Percentage of occurrence')
     plt.title('Frequency of alert category')
     if SURICATA_SUMMARY:
-        plt.xticks(ind, ('c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9', 'c10', 'c11', 'c12', 'c13', 'c14'))
+        plt.xticks(indices, ('c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9', 'c10', 'c11', 'c12', 'c13', 'c14'))
     else:
-        plt.xticks(ind, [x.split('.')[1] for x in vals], rotation='vertical')
+        plt.xticks(indices, [x.split('.')[1] for x in micro_attack_stages], rotation='vertical')
     plt.tick_params(axis='x', which='major', labelsize=8)
     plt.tick_params(axis='x', which='minor', labelsize=8)
-    #plt.yticks(np.arange(0, 13000, 1000))
-    plt.legend([plot[0] for plot in p], team_labels)
+    # plt.yticks(np.arange(0, 13000, 1000))
+    plt.legend([plot[0] for plot in plots], team_labels)
     plt.tight_layout()
-    #plt.show() 
+    # plt.show()
     return plt
 
+
 ## 14
 def legend_without_duplicate_labels(ax, fontsize=10, loc='upper right'):
     handles, labels = ax.get_legend_handles_labels()
     unique = [(h, l) for i, (h, l) in enumerate(zip(handles, labels)) if l not in labels[:i]]
-    unique = sorted(unique, key = lambda x: x[1])
-    ax.legend(*zip(*unique), loc=loc, fontsize=fontsize) 
-    
+    unique = sorted(unique, key=lambda x: x[1])
+    ax.legend(*zip(*unique), loc=loc, fontsize=fontsize)
+
+
+def _get_ups_and_downs(frequencies, slopes):
+    positive = [(0, slopes[0])]  # (index, slope)
+    positive.extend([(i, slopes[i]) for i in range(1, len(slopes)) if (slopes[i - 1] <= 0 and slopes[i] > 0)])
+    negative = [(i + 1, slopes[i + 1]) for i in range(0, len(slopes) - 1) if (slopes[i] < 0 and slopes[i + 1] >= 0)]
+    if slopes[-1] < 0:  # Special case for last ramp down that's not fully gone down
+        negative.append((len(slopes), slopes[-1]))
+    elif slopes[-1] > 0:  # Special case for last ramp up without any ramp down
+        negative.append((len(slopes), slopes[-1]))
+
+    common = set(negative).intersection(positive)
+    negative = [item for item in negative if item not in common]
+    positive = [item for item in positive if item not in common]
+
+    negative = [x for x in negative if (frequencies[x[0]] <= 0 or x[0] == len(frequencies) - 1)]
```

Comment:
I was wondering what is the purpose of these two lines (originally, `negative = [x for x in negative if (y[x[0]] <= 0 or x[0] == len(y)-1)]` and `positive = [x for x in positive if (y[x[0]] <= 0 or x[0] == 0)]`, same functionality)? `y` or `frequencies` cannot be negative. I was looking at these lines, but could not really understand what they are for

Reply:
positive has all positive slopes and negative has all negative slopes.

Reply:
I see. Thanks for the explanation. Would it be better to use `==` instead of `<=`? The `<` part was the one that confused me, since a length of an episode cannot be negative